### PR TITLE
fix: add ReDoS protection to regex pattern validation

### LINKED ISSUE #521: [Security] Backend: ReDoS vulnerability in regex pattern validation
## Summary

A Regular Expression Denial of Service (ReDoS) vulnerability exists in the `_check_username_list_regex()` function in `buffalogs/impossible_travel/modules/alert_filter.py`. 

The function compiles user-provided regex patterns without validation, allowing malicious patterns to cause CPU exhaustion and application unavailability.

### Severity
**High** - Can lead to Denial of Service

### Affected Component
- **File:** `buffalogs/impossible_travel/modules/alert_filter.py`
- **Function:** `_check_username_list_regex()` (lines 106-117)
- **Attack Vector:** `Config.ignored_users`, `Config.enabled_users`, `Config.vip_users` fields

### Current Vulnerable Code

```python
def _check_username_list_regex(word: str, values_list: list) -> bool:
    for item in values_list:
        if word == item:
            return True
        try:
            regexp = re.compile(item)  # ⚠️ NO VALIDATION
            if regexp.search(word):
                return True
        except re.error:
            continue
    return False
Proof of Concept
An attacker with admin access could add a malicious regex pattern:

# Malicious pattern
Config.ignored_users = [r"(a+)+"]

# Trigger with crafted input
username = "a" * 50 + "X"

# Result: CPU hangs for minutes/hours due to catastrophic backtracking
Dangerous patterns that can cause ReDoS:
(a+)+ - Nested quantifiers
(a*)* - Nested star operators
(a|ab)* - Overlapping alternation
(\w+)+b - Exponential backtracking

### PR DESCRIPTION
**Pull request by kami922**
Implements comprehensive protection against Regular Expression Denial of Service (ReDoS) attacks in the username filtering system.

## Problem
The _check_username_list_regex() function compiled user-provided regex patterns without validation, allowing malicious patterns like (a+)+ to cause exponential-time execution and CPU exhaustion. This could be exploited via Config.ignored_users, Config.enabled_users, or Config.vip_users fields to create a Denial of Service condition.

## Solution
Added multi-layer validation in new _is_safe_regex() function:

1. Length validation: Reject patterns > 100 characters
2. Complexity check: Reject patterns with > 50 special regex chars
3. Catastrophic backtracking detection: Identify nested quantifiers
   - Patterns like (a+)+, (a*)*, (a|ab)* are detected and rejected
4. Syntax validation: Ensure valid regex syntax before compilation

Updated _check_username_list_regex() to validate all patterns before compilation. Unsafe patterns are logged and skipped gracefully.

## Testing
- Added 15 comprehensive tests in TestReDoSProtection class
- All 35 tests pass (15 new + 20 existing)
- Tests cover: safe patterns, dangerous patterns, edge cases, performance, and full integration scenarios

## Security Impact
- Prevents ReDoS attacks that could hang the application
- Maintains backward compatibility with safe regex patterns
- Minimal performance overhead (< 1ms per pattern validation)
- Comprehensive logging for security monitoring

## Files Changed
- buffalogs/impossible_travel/modules/alert_filter.py: Added validation
- buffalogs/impossible_travel/tests/detection/test_alert_filter.py: Added tests
- REDOS_FIX_DOCUMENTATION.md: Complete technical documentation

### CODE REVIEW

**Comment by kami922** [comment]
@Lorygold  Hello
1- Yes, I think  it would be a better approach. Moving the _is_safe_regex() check to Config.clean() ensures we validate once at the entry point,ensuring we  validate once at the entry point. I shall implement it in next commit
2-  To handle legacy data, I will create a management command to scan and log any existing dangerous patterns for admin review. The alert filter will retain a simple try-except as a safety fallback, but the primary validation burden will shift to the config level

let me know if it sounds reasonable to you so i will implement it

**Comment by Lorygold** [comment]
No problem! It’s marked as a draft simply because it’s not finished yet, so it serves as a reminder that it shouldn’t be merged for now

**Comment by kami922** [comment]
1. Validation at Config Save Time

Added [validate_regex_patterns()](vscode-file://vscode-app/snap/code/218/usr/share/code/resources/app/out/vs/code/electron-browser/workbench/workbench.html) validator to [validators.py](vscode-file://vscode-app/snap/code/218/usr/share/code/resources/app/out/vs/code/electron-browser/workbench/workbench.html) that uses [_is_safe_regex()](vscode-file://vscode-app/snap/code/218/usr/share/code/resources/app/out/vs/code/electron-browser/workbench/workbench.html) to check patterns. This validator is now applied to the [ignored_users](vscode-file://vscode-app/snap/code/218/usr/share/code/resources/app/out/vs/code/electron-browser/workbench/workbench.html), [enabled_users](vscode-file://vscode-app/snap/code/218/usr/share/code/resources/app/out/vs/code/electron-browser/workbench/workbench.html), and [vip_users](vscode-file://vscode-app/snap/code/218/usr/share/code/resources/app/out/vs/code/electron-browser/workbench/workbench.html) fields in the Config model.

Result: When admins try to save configs with unsafe patterns via the Django admin panel, they'll get a clear ValidationError explaining why the pattern was rejected. No repeated validation during alert processing.

2. Check Existing Configs in Migration 0022

Added [check_regex_patterns()](vscode-file://vscode-app/snap/code/218/usr/share/code/resources/app/out/vs/code/electron-browser/workbench/workbench.html) function to migration 0022 following the same pattern as [populate_device_fingerprint()](vscode-file://vscode-app/snap/code/218/usr/share/code/resources/app/out/vs/code/electron-browser/workbench/workbench.html). The function:

Iterates through all existing Config entries
Checks for unsafe patterns in the three user list fields
Logs warnings with details about which patterns are unsafe
Non-blocking: Doesn't prevent migration from completing
Added to operations:
migrations.RunPython(check_regex_patterns, migrations.RunPython.noop),

**Comment by Lorygold** [comment]
why all those files have been changed? some problems with the linters version locally and in the CI?

**Comment by kami922** [comment]
@Lorygold  TBH I  could not really figure it out I did apply pre commit hook and it said the linting tests passed locally, but failed on github actions.
I applied Black formatting multiple times before committing and pushing  and it passed locally but for some reason it failed online adjusted versions still same problem. CI enforces Black (line-length 160) and isort, so once we aligned with the repo’s configs the formatter touched many files. The functional change is small—moving _is_safe_regex into validators to break a circular import—everything else is formatting to satisfy the existing linting rules.

File: .github/.pre-commit-config.yaml [hunk 1]
```diff
@@ -13,7 +13,7 @@ repos:
         args: ["--settings-path", ".github/configurations/python_linters/.isort.cfg", "--filter-files", "--skip", "venv"]
 
 -   repo: https://github.com/psf/black
-    rev: 24.8.0
+    rev: 25.1.0
```
Comment (Lorygold) [inline .github/.pre-commit-config.yaml]: maybe because you change this version?

**Comment by Lorygold** [comment]
you should apply changes only to the files needed for your features, not all those `81 files`

**Comment by kami922** [comment]
@Lorygold  Hello I have made a clean commit and hope it works
now regarding the issues associated with formatters I have observed some inconsistencies among versions

### Current Versions

**Pre-commit config** (`.github/.pre-commit-config.yaml`):
- Black: `24.8.0`
- isort: `5.13.2`
- flake8: `7.1.1`

**CI workflow** (`.github/workflows/_python.yml`):
- Black: `25.1.0`
- isort: `6.0.1`
- flake8: `7.1.2`


I think it should be consistent  also I would recommend shifting to ruff I am currently working on this issue in intel owl. however it can be a time consuming issue requiring discipline and also would require changing in all files.

**Comment by Lorygold** [comment]
If you aligned your branch with the `develop` one, it should be already fixed, but feel free to let me know if the error is still there and I missed something.
The problem persists only for the `buffacli_CI` that I'll update it now

**Comment by kami922** [comment]
@Lorygold Hello ok thats great to know let me know whats left for pr to be closed and merged

File: buffalogs/impossible_travel/migrations/0022_remove_tasksettings_unique_task_execution_mode_and_more.py [hunk 2]
```diff
@@ -81,6 +81,31 @@ def populate_device_fingerprint(apps, schema_editor):
...
 
+def check_regex_patterns(apps, schema_editor):
+    """Check existing Config entries for unsafe regex patterns and log warnings.
+
+    This is a non-blocking check that identifies existing dangerous patterns
+    in ignored_users, enabled_users, and vip_users fields. Admins should
+    review and update these patterns in the Django admin panel.
+    """
+    Config = apps.get_model("impossible_travel", "Config")
+    from impossible_travel.validators import _is_safe_regex
```
Comment (Lorygold) [inline buffalogs/impossible_travel/migrations/0022_remove_tasksettings_unique_task_execution_mode_and_more.py:92]: please define all the imports at the beginning of the file

**Comment by Lorygold** [comment]
also @Noble-47 will give a check soon

**Comment by kami922** [comment]
@Lorygold  Hello  any update on this one?

**Comment by Noble-47** [comment]
Hi @kami922 

I'll give you an update soon.

**Comment by kami922** [comment]
@Noble-47  Hello just dropping by a reminder. Also please do let me know if there is anything i could do to speed up the review process.
